pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/GateImplementationsLM.hpp

```diff
@@ -490,17 +490,23 @@ class GateImplementationsLM : public PauliGenerator<GateImplementationsLM> {
             Pennylane::Util::revWireParity(rev_wires);
         PL_ASSERT(nw_tot == parity.size() - 1);
 
-        const std::size_t dim = one << n_wires;
+        const std::vector<std::size_t> indices =
+            (n_contr == 0) ? Gates::generateBitPatterns(wires, num_qubits)
+                           : std::vector<std::size_t>{};
 
         PL_LOOP_PARALLEL(1)
         for (std::size_t k = 0; k < exp2(num_qubits - nw_tot); k++) {
-            const auto indices =
-                parity2indices(k, parity, rev_wire_shifts, n_contr, rev_wires);
-            std::vector<std::complex<PrecisionT>> coeffs_in(dim);
-            for (std::size_t i = 0; i < dim; i++) {
-                coeffs_in[i] = arr[indices[i]];
+            std::size_t offset{0U};
+            if (n_contr == 0) {
```

Comment:
I don't think it matters here since `n_contr` is `const`, but ok.

Reply:
It is important because `n_contr` is not a `constexpr`. Therefore, each iteration will evaluate the same condition on runtime.

Reply:
This is true for a debug build, but for an optimized build, the following (for example)
```
void f(std::complex<double> * arr, std::size_t num){
    const bool skip = false;
    std::vector<double> c(num);
    std::vector<double> d(num);
    std::vector<double> v(num);
    if (skip){
        for (std::size_t i=0; i < num; i++){
            arr[i] = c[i] + d[i] * v[i];
        }
    }else{
        for (std::size_t i=0; i < num; i++){
                arr[i] = c[i] * d[i] + v[i];
        }
    }
}
```
results in the same assembly as 
```
void f(std::complex<double> * arr, std::size_t num){
    const bool skip = false;
    std::vector<double> c(num);
    std::vector<double> d(num);
    std::vector<double> v(num);
    for (std::size_t i=0; i < num; i++){
        if (skip){
            arr[i] = c[i] + d[i] * v[i];
        }else{
            arr[i] = c[i] * d[i] + v[i];
        }
    }
}
```
So here I'll use the first one because we do not save much code, but otherwise it might be worth using the second option.